x/transferrouter/abci/proposal.go

```
@@ -0,0 +1,324 @@
+	TxConfig   client.TxConfig
+}
+
+func NewProposalHandler(opts ProposalHandlerOptions) *ProposalHandler {
+	return &ProposalHandler{
+		keeper:     opts.Keeper,
+		txSelector: opts.TxSelector,
+		signer:     opts.Signer,
+		txVerifier: opts.TxVerifier,
+		txConfig:   opts.TxConfig,
+	}
+}
+
+var CallMaxGas = uint64(1000000) // arbitrary value
+
+func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		logger := h.keeper.Logger(ctx)
+
+		// 1. Add the source callback queue
+		chainId, err := utils.ParseChainID(ctx.ChainID())
+		if err != nil {
+			logger.Error("failed to parse chain id", "error", err)
+			return nil, errors.New("failed to parse chain id")
+		}
+
+		var maxBlockGas uint64
+		consParams := ctx.ConsensusParams()
+		if consParams.Block != nil {
+			maxBlockGas = uint64(consParams.Block.MaxGas)
+		}
+
+		if h.txSelector == nil {
+			logger.Error("tx selector is nil")
+			return nil, errors.New("tx selector is nil")
+		}
+
+		defer h.txSelector.Clear()
+
+		params, err := h.keeper.Params.Get(ctx)
+		if err != nil {
+			logger.Error("Failed to get params", "error", err)
+			return nil, errors.New("failed to get params")
+		}
+
+		// Parse the configured private key (in hex format) and derive the corresponding
+		// Ethereum address of the known signer.
+		if params.KnownSignerPrivateKey == "" {
+			logger.Error("known signer private key is empty")
+			return nil, errors.New("known signer private key is empty")
+		}
+		privKey, err := crypto.HexToECDSA(params.KnownSignerPrivateKey)
+		if err != nil {
+			logger.Error("failed to parse private key", "error", err)
+			return nil, errors.New("failed to parse private key")
+		}
+
+		knownSignerBz := crypto.PubkeyToAddress(privKey.PublicKey).Bytes()
+		nextNonce, err := h.keeper.AccountKeeper.GetSequence(ctx, sdk.AccAddress(knownSignerBz))
+		if err != nil {
+			nextNonce = 0
```

We should throw an error here, and instead always initialize the account in InitGenesis

---

sounds good, will do that